Add Guide for Creating Custom Notes (#21)
0xPolygonMiden/miden-tutorials: docs/rust-client/custom_note_how_to.md

> # How To Create a Custom Note on Miden

Reviewer: I think that the formulation of the title is missleading. There are no standard or custom notes. They are all just notes imo. So it would more be "How To Create a Note Script on Miden" or "How to Create a Note".

Author: Will fix.

*Creating custom notes*

## Overview

In this guide, we will create a custom note on Miden that can only be consumed by someone who knows the preimage of the hash stored in the note. This approach securely embeds assets into the note and restricts spending to those who possess the correct secret number.

By following the steps below and using the Miden Assembly code and Rust example, you will learn how to:

> * Create a note with custom logic.

Reviewer: Can you make sure that your markdown formatting is applied in the same way throughout, e.g. here you use * when in other places you use - for bullet points.

Author: Will fix.

* Leverage Miden’s privacy features to keep certain transaction details private.

> Unlike Ethereum, where all transaction details are publicly visible in the mempool, Miden enables you to partially or completely hide transaction details.

Reviewer: I would remove "in the mempool" because everything is public always.

Author: I think we should keep the reference to the "mempool" because specifically this exact type of application would not be possible on ethereum exactly because the mempool is public, and the tx would be front run.

## What we'll cover

> - Creating custom note logic

Reviewer: Imo it's more "Writing Miden Assembly"

Reviewer: "Writing Miden Assembly for a Note Script"

Author: Fixed.

    # Will fail if the two hashes do not match
    assert_eqw
    # => []
    
    # Write the asset in note to memory address 0
    push.0 exec.note::get_assets
    # => [num_assets, dest_ptr]
    
    drop
    # => [dest_ptr]
    
    # Load asset from memory
    mem_loadw
    # => [ASSET]
    
    # Call receive asset in wallet
    call.wallet::receive_asset
    # => []
end
```

### How the Assembly Code Works:

1. **Passing the Secret:**  
   The secret number is passed as `Note Arguments` into the note.
   
2. **Hashing the Secret:**  
   The `hperm` instruction applies a hash permutation to the secret number, resulting in a hash that takes up four stack elements.
   
3. **Stack Cleanup and Comparison:**  
   The assembly code extracts the digest, loads the note inputs from memory and checks if the computed hash matches the note’s stored hash.
   
4. **Asset Transfer:**  
   If the hash of the number passed in as `Note Arguments` matches the hash stored in the note inputs, the script continues, and the asset stored in the note is loaded from memory and passed to Bob’s wallet via the `wallet::receive_asset` function.

### 5. Consuming the Note

> With the note created, Bob can now consume it—but only if he provides the correct secret. When Bob initiates the transaction to consume the note, he must supply the same secret number used when Alice created the note. The custom note’s logic will hash the secret and compare it with its stored hash. If they match, Bob’s wallet receives the asset.

Reviewer: ```suggestion
The note having been created, Bob can now consume it if he provides the correct secret. When Bob initiates the transaction to consume the note, he must supply the same secret number used when Alice created the note. The custom note’s logic will hash the secret and compare it with its stored hash. If they match, Bob’s wallet receives the asset.
```

Author: "having been" is passive voice and sounds less active: ["Passive Voice: When to Use It and When to Avoid It"](https://www.grammarly.com/blog/grammar/passive-voice/#:~:text=Updated%20on%20December%2019%2C%202024,past%20participle%20of%20the%20verb.)

    let (account, seed) = builder.build().unwrap();
    client
        .add_account(
            &account,
            Some(seed),
            &AuthSecretKey::RpoFalcon512(key_pair),
            false,
        )
        .await?;
    Ok(account)
}

#[tokio::main]
async fn main() -> Result<(), ClientError> {
    // Initialize client
    let mut client = initialize_client().await?;
    println!("Client initialized successfully.");

    // Fetch latest block from node
    let sync_summary = client.sync_state().await.unwrap();
    println!("Latest block: {}", sync_summary.block_num);

    //------------------------------------------------------------
    // STEP 1: Create two basic accounts (Alice and Bob)
    //------------------------------------------------------------
    println!("\n[STEP 1] Creating new accounts");
    let alice_account = create_basic_account(&mut client).await?;
    println!("Alice's account ID: {:?}", alice_account.id().to_hex());
    let bob_account = create_basic_account(&mut client).await?;
    println!("Bob's account ID: {:?}", bob_account.id().to_hex());

    //------------------------------------------------------------

>     // STEP 2: Mint and consume tokens for Alice with Ephemeral P2ID

Reviewer 2: I think we can delete the mentioning of ephemeral P2ID here, no? Let's use just normal public notes to let the user focus on the purpose of the tutorial "creating custom notes". I would make everything else as easy as possible

Author: Made the "how to" not use ephemeral notes

    //------------------------------------------------------------
    println!("\n[STEP 2] Mint tokens with Ephemeral P2ID");

0xf42ab06ffd227c2000005b2767bc5e    let faucet_id = AccountId::from_hex("0xf42ab06ffd227c2000005b2767bc5e").unwrap();
    let amount: u64 = 100;
    let fungible_asset_mint_amount = FungibleAsset::new(faucet_id, amount).unwrap();

    let transaction_request = TransactionRequestBuilder::mint_fungible_asset(
        fungible_asset_mint_amount.clone(),
        alice_account.id(),
        NoteType::Public,
        client.rng(),
    )
    .unwrap()
    .build();

    let tx_execution_result = client
        .new_transaction(faucet_id, transaction_request)
        .await?;
    client
        .submit_transaction(tx_execution_result.clone())
        .await?;

    // The minted fungible asset is public so output is a `Full` note type
    let p2id_note: Note =
        if let OutputNote::Full(note) = tx_execution_result.created_notes().get_note(0) {
            note.clone()
        } else {

>             panic!("Expected Full note type");

Reviewer 2: Can you say `panic!("Expected OutputNote::Full")?

Author: Fixed.

        };

    let transaction_request = TransactionRequestBuilder::new()
        .with_unauthenticated_input_notes([(p2id_note, None)])
        .build();

    let tx_execution_result = client
        .new_transaction(alice_account.id(), transaction_request)
        .await?;
    client.submit_transaction(tx_execution_result).await?;
    client.sync_state().await?;

    // -------------------------------------------------------------------------
    // STEP 3: Hash Secret Number and Build Note
    // -------------------------------------------------------------------------

>     println!("\n[STEP 3] Create note");

Reviewer 2: Can you say here "Create custom note" to make is super clear that this is what the tutorial is about

    // Prepend an empty word (4 zero Felts) for the RPO
    note_secret_number.splice(0..0, Word::default().iter().cloned());
    let secret_number_digest = Hasher::hash_elements(&note_secret_number);
    println!("digest: {:?}", secret_number_digest);

    let assembler: Assembler = TransactionKernel::assembler().with_debug_mode(true);
    let file_path = Path::new("../masm/notes/hash_preimage_note.masm");
    let code = fs::read_to_string(file_path).unwrap();
    let rng = client.rng();
    let serial_num = rng.draw_word();
    let note_script = NoteScript::compile(code, assembler).unwrap();

    let inputs: [Felt; 4] = secret_number_digest.into();
    let note_inputs = NoteInputs::new(inputs.into()).unwrap();

    let recipient = NoteRecipient::new(serial_num, note_script, note_inputs);
    let tag = NoteTag::for_public_use_case(0, 0, NoteExecutionMode::Local).unwrap();
    let aux = Felt::new(0);
    let metadata = NoteMetadata::new(
        alice_account.id(),
        NoteType::Public,
        tag,
        NoteExecutionHint::always(),
        aux,
    )?;
    let vault = NoteAssets::new(vec![fungible_asset_mint_amount.clone().into()])?;

>     let increment_note = Note::new(vault, metadata, recipient);

Reviewer 2: What is the `increment_note`. Do we need another name for it? Maybe `custom_note`?

Author: this was a typo, will fix

    println!("note hash: {:?}", increment_note.hash());

    let output_note = OutputNote::Full(increment_note.clone());
    let incr_note_create_request = TransactionRequestBuilder::new()
        .with_own_output_notes([output_note].to_vec())
        .unwrap()
        .build();

    let tx_result = client
        .new_transaction(alice_account.id(), incr_note_create_request)
        .await
        .unwrap();
    let tx_id = tx_result.executed_transaction().id();
    println!(
        "View transaction on MidenScan: https://testnet.midenscan.com/tx/{:?}",
        tx_id
    );
    let _ = client.submit_transaction(tx_result).await;
    client.sync_state().await.unwrap();

    // -------------------------------------------------------------------------

>     // STEP 4: Consume Note

Reviewer 2: Can you say here "consume the custom note"?

Author: Fixed.

    // -------------------------------------------------------------------------

>     println!("\n[STEP 4] Bob consumes the Ephemeral Hash Preimage Note with Correct Secret");

Reviewer 2: Let's not make the custom note consumption ephemeral

Author: Will fix

Author: No longer using ephemeral notes.

# How To Create a Custom Note on Miden

*Creating custom notes*

## Overview

In this guide, we will create a custom note on Miden that can only be consumed by someone who knows the preimage of the hash stored in the note. This approach securely embeds assets into the note and restricts spending to those who possess the correct secret number.

By following the steps below and using the Miden Assembly code and Rust example, you will learn how to:

* Mint and store assets within a custom note.
* Require a secret preimage to consume the note.
* Leverage Miden’s privacy features to keep transaction details private.

Unlike Ethereum, where transaction details are publicly visible in the mempool, Miden enables you to partially or completely hide transaction details.

## What we'll cover

- Creating custom note logic
- Consuming notes

## Step-by-Step Process

### 1. Creating Two Accounts: Alice & Bob

First, we create two basic accounts for the two users:
- **Alice:** The account that creates and funds the custom note.
- **Bob:** The account that will consume the note if he knows the secret.

> ### 2. Deploying a Fungible Faucet

Author: Yes, this is a good idea. Will fix.

> ### 2. Deploying a Fungible Faucet

Author: The only issue is that the faucet account must be present in the miden-client. More context here: https://github.com/0xPolygonMiden/miden-client/issues/762

Next, we deploy a fungible faucet. This faucet is used to mint assets (tokens) that will be included in the note. The asset is identified by a token symbol (e.g., MID) and minted to Alice’s account. This ensures that Alice has the necessary tokens to create the note.

### 3. Hashing the Secret Number

The security of the custom note hinges on a secret number. Here, we will:

> - Choose a secret number (for example, an array of four field elements).

Reviewer 2: Can you just say integers instead of `field elements`?

I would try to abstract all zk-jargon away

- For simplicity, we're only hashing 4 elements. Therefore, we prepend an empty word—consisting of 4 zero Felts—as a placeholder. This is required by the RPO hashing algorithm to ensure the input has the correct structure and length for proper processing.
- Compute the hash of this secret. The resulting hash will serve as the note’s input, meaning that the note can only be consumed if the secret number’s hash preimage is provided during consumption.

### 4. Creating the Custom Note

Now, combine the minted asset and the secret hash to build your custom note. The note is created using the following key steps:

1. **Note Inputs:**  
   - The note is set up with the asset and the hash of the secret number.
   
2. **Miden Assembly Code:**  
   - The custom Miden Assembly script ensures that the note can only be consumed if the provided secret, when hashed, matches the original hash stored in the note.

Below is the Miden Assembly code for the note:

```masm
use.miden::note
use.miden::contracts::wallets::basic->wallet

# => [HASH_PREIMAGE_SECRET]
begin

>     # => [HASH_PREIMAGE_SECRET]

Author: Fixed.

    hperm

    # => [F,E,D]
    # E is digest
    dropw swapw dropw

    # => [DIGEST]
    # writing note inputs to memory
    push.0 exec.note::get_inputs drop drop

    # => [DIGEST]
    # pad stack and load note inputs from memory
    padw push.0 mem_loadw

    # => [INPUTS, DIGEST]
    # assert equality
    assert_eqw

    # => []
    # write assets in note to memory address 0
    push.0 exec.note::get_assets

    # => [num_assets, dest_ptr]
    drop

    # => [dest_ptr]
    # load asset from memory
    mem_loadw

    # => [ASSET]
    # call receive asset in wallet
    call.wallet::receive_asset

end
```

### How the Assembly Code Works:

1. **Passing the Secret:**  

>    The secret number is passed as an argument into the note.

Reviewer 2: Can you refer here explicitly to `Note arguments` and maybe even link to the part in the docs?

The following Rust code demonstrates how to implement the steps outlined above using the Miden client library:

```rust
use std::{fs, path::Path, sync::Arc};

use rand::Rng;
use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};

use miden_client::{
    account::{
        component::{BasicFungibleFaucet, BasicWallet, RpoFalcon512},
        AccountBuilder, AccountStorageMode, AccountType,
    },
    asset::{FungibleAsset, TokenSymbol},
    crypto::RpoRandomCoin,
    note::{
        Note, NoteAssets, NoteExecutionHint, NoteExecutionMode, NoteInputs, NoteMetadata,
        NoteRecipient, NoteScript, NoteTag, NoteType,
    },
    rpc::{Endpoint, TonicRpcClient},
    store::{sqlite_store::SqliteStore, StoreAuthenticator},
    transaction::{OutputNote, TransactionKernel, TransactionRequestBuilder},
    Client, ClientError, Felt,
};

use miden_crypto::{hash::rpo::Rpo256 as Hasher, rand::FeltRng};

use miden_objects::{
    account::AuthSecretKey, assembly::Assembler, crypto::dsa::rpo_falcon512::SecretKey, Word,
};

// Initialize client helper
pub async fn initialize_client() -> Result<Client<RpoRandomCoin>, ClientError> {
    let endpoint = Endpoint::new(
        "https".to_string(),

>         "rpc.testnet.miden.io".to_string(),

Author: Yes will look into minting from a faucet address.

>         "rpc.testnet.miden.io".to_string(),

Author: To mint using the public faucet address is not exactly straight forward right now using the Miden client 0.7. On the `next` branch there is a new function that allows you to import an account by id. This would make it super easy to mint tokens from the faucet. 

https://github.com/0xPolygonMiden/miden-client/issues/722

>         "rpc.testnet.miden.io".to_string(),

Reviewer 2: OK, then merge the example as is for now and we will revisit it once we released v0.8
